core/src/main/scala/kafka/admin/AdminUtils.scala

```diff
@@ -106,37 +106,43 @@ object AdminUtils extends Logging {
                               replicationFactor: Int,
                               fixedStartIndex: Int = -1,
                               startPartitionId: Int = -1,
-                              rackInfo: Map[Int, String] = Map())
-  : Map[Int, Seq[Int]] = {
+                              rackInfo: Map[Int, String] = Map()): Map[Int, Seq[Int]] = {
     if (nPartitions <= 0)
       throw new AdminOperationException("number of partitions must be larger than 0")
     if (replicationFactor <= 0)
       throw new AdminOperationException("replication factor must be larger than 0")
     if (replicationFactor > brokerList.size)
-      throw new AdminOperationException("replication factor: " + replicationFactor +
-        " larger than available brokers: " + brokerList.size)
-    if (rackInfo.size > 0) {
-      val brokerRackMap = rackInfo.filterKeys(brokerList.contains(_))
-      if (brokerRackMap.size < brokerList.size) {
-        throw new AdminOperationException("Not all brokers have rack information for replica rack aware assignment");
-      }
-      return assignReplicasToBrokersRackAware(nPartitions, replicationFactor, brokerRackMap, fixedStartIndex,
+      throw new AdminOperationException(s"replication factor: $replicationFactor larger than available brokers: ${brokerList.size}")
+    if (rackInfo.isEmpty)
+      assignReplicasToBrokersRackUnaware(nPartitions, replicationFactor, brokerList, fixedStartIndex, startPartitionId)
+    else {
+      val brokerRackMap = rackInfo.filterKeys(brokerList.contains).toMap
+      if (brokerRackMap.size < brokerList.size)
+        throw new AdminOperationException("Not all brokers have rack information for replica rack aware assignment")
+      assignReplicasToBrokersRackAware(nPartitions, replicationFactor, brokerRackMap, fixedStartIndex,
         startPartitionId)
     }
-    val ret = new mutable.HashMap[Int, List[Int]]()
-    val startIndex = if (fixedStartIndex >= 0) fixedStartIndex else rand.nextInt(brokerList.size)
-    var currentPartitionId = if (startPartitionId >= 0) startPartitionId else 0
+  }
 
-    var nextReplicaShift = if (fixedStartIndex >= 0) fixedStartIndex else rand.nextInt(brokerList.size)
+  private def assignReplicasToBrokersRackUnaware(nPartitions: Int,
+                                                 replicationFactor: Int,
+                                                 brokerList: Seq[Int],
+                                                 fixedStartIndex: Int = -1,
+                                                 startPartitionId: Int = -1): Map[Int, Seq[Int]] = {
+    val ret = mutable.Map[Int, Seq[Int]]()
+    val brokerArray = brokerList.toArray
```

Comment:
We are receiving a `Seq` (which could be a `List`) but we are accessing it via an index. If the received implementation is a `List`, then each access is O(n). There are a couple of ways to fix this: change the type to `IndexedSeq` which would affect callers or just convert to an IndexedSeq. For the usage we have here, `Array` seemed like the appropriate type.